# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [ ]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [80]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

dataset = pd.read_csv('science_data_large.csv')
print(dataset.head(15))
print(dataset.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [81]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataset[['Temperature °C','Mols KCL']]
y = dataset['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=42)
# For grading consistency use random_state=42 
# Output the result to show it worked
print("Features DataFrame (X_train):")
print(X_train.head())  # Display the first few rows of the training features
print("\nLabels Series (y_train):")
print(y_train.head())  # Display the first few rows of the training labels

print("\nFeatures DataFrame (X_test):")
print(X_test.head())  # Display the first few rows of the test features
print("\nLabels Series (y_test):")
print(y_test.head())  # Display the first few rows of the test labels

Features DataFrame (X_train):
     Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421

Labels Series (y_train):
716    103064.31430
351    193753.02860
936     21670.02857
256    505027.40000
635     75092.02857
Name: Size nm^3, dtype: float64

Features DataFrame (X_test):
     Temperature °C  Mols KCL
521             100       541
737             635       668
740             799       665
660             966       871
411             785       595

Labels Series (y_test):
521    1.177623e+05
737    8.687293e+05
740    1.084893e+06
660    1.716039e+06
411    9.536850e+05
Name: Size nm^3, dtype: float64


## 3. Perform a Linear Regression

In [82]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
# Create a sample datapoint as a DataFrame to avoid warning
sample = pd.DataFrame([[300, 400]], columns=['Temperature °C', 'Mols KCL'])
predicted_size = model.predict(sample)
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
# Display the results
print(f"Predicted size for the sample datapoint {sample.values.tolist()}: {predicted_size[0]}")
print(f"Model Score: {score}")
print(f"Coefficients: {coefficients}, Intercept: {intercept}")

Predicted size for the sample datapoint [[300, 400]]: 263530.47102248715
Model Score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649], Intercept: -409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)

$$h(x) = Intercept + (Coefficient for Temperature \cdot Temperature °C) + (Coefficient for Mols KCL \cdot Mols KCL)$$

$$
h(x) = -409391.48 + (866.15 \cdot Temperature) + (1032.70 \cdot Mols KCL) 
$$

Report on score and explain meaning:

The reported score for this model is 0.86. Now the question lies is how do we make sense of this number. What the score calculated describes is the coefficient of determination. This coefficient of determination is a measure of how well the independent variables, or features, explain the variability of the dependent variable, or the label (size of slime). A value of 0.86 means that approximately 86% of the variability in the size of the slime can be explained by the linear regression model using the temperature and mols of KCl as predictors. This indicates a strong relationship between the predictors and the response variable.
Conversely, the remaining 14% of the variability is explained by factors not included in the model or random chance. A higher $R^2$ value, closer to 1, suggests that the model fits the data well, whereas a lower value indicates that the model does not explain much of the variability in the response.

## 4. Use Cross Validation

In [83]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kfold = KFold(n_splits=5, random_state=42, shuffle=True)
scores = cross_val_score(model, X, y, cv=kfold)

# Print the scores for each fold
print("Cross-validation scores for each fold:", scores)

# Calculate mean R² score
mean_score = scores.mean()
# Calculate standard deviation
std_dev = scores.std()

print("Mean R² score:", mean_score)
print("Standard deviation of R² scores:", std_dev)
# Report on their finding and their significance

Cross-validation scores for each fold: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean R² score: 0.8597294202684646
Standard deviation of R² scores: 0.01838773713930639


Write findings here: From this part, what we can conclude from our results is that from our calculated cross-validation scores, we receive a mean of 0.86. This mean score means that approximately 86% of the variance in the size of the slime can be explained by the model. This indicated that the model effectively captured some patterns of the data, acting as a good fit for a model. As for our standard deviation, we have a value reported of 0.01. This small value signifies the variability of the model across different sets of data. A value closer to 0 would be indicative that the model would perform consistently across folds and gives us confidence of the reliability of the model's predictions. Since our value in this case is 0.01, very close to 0, we can conclude that this model would be reliable across different data sets. 

## 5. Using Polynomial Regression

In [84]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
# Report on the metrics and output the resultant equation as you did in Part 3.


poly = PolynomialFeatures(degree=2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

# Initialize the polynomial regression model
poly_model = LinearRegression()

# Train the model on the training set
poly_model.fit(X_train_poly, y_train)

# Evaluate the model on the test set
poly_score = poly_model.score(X_test_poly, y_test)

# Extract coefficients and intercept
poly_coefficients = poly_model.coef_
poly_intercept = poly_model.intercept_

# Display results
print(f"Model Score: {poly_score}")
print(f"Coefficients: {poly_coefficients}, Intercept: {poly_intercept}")


Model Score: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27197088e-07  1.26458843e-11
  2.00000000e+00  2.85714287e-02], Intercept: 2.0479375962167978e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

We can use the following polynomial equation:
$$
h(x) = a + (b_1 \cdot \text{Temperature} + b_2 \cdot \text{Mols KCL} + c_1 \cdot \text{Temperature}^2 + c_2 \cdot \text{Mols KCL}^2 + d \cdot \text{Temperature} \cdot \text{Mols KCL})
$$
Then by plugging in our values we have:
$$
h(x) = 2.05 + (12.00 \cdot \text{Temperature} + -1.27e{-07} \cdot \text{Mols KCL} + 1.26e{-11} \cdot \text{Temperature}^2 + 2.00 \cdot \text{Mols KCL}^2 + 2.86e{-02} \cdot \text{Temperature} \cdot \text{Mols KCL})
$$

Report on the score and interpret: 
The value of our score being 1.0 suggests that the model has perfectly fit the data and accounts for 100% of the variance in the target variable based on the features. This would imply that the model is reliable for the test set, however, it may be overfitted and not be able to generalise well given other datasets. This could mean that the model may have memorised the training dataset to make predictions instead of learning the general trends. 

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [85]:
# Load the dataset. Then train and evaluate the classification models.


ckd_dataset = pd.read_csv('ckd_feature_subset.csv')

X = ckd_dataset[["age","bp","wbcc","appet_poor","appet_good","rbcc"]]
y = ckd_dataset["Target_ckd"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=42)

kfold  = KFold(n_splits=5, random_state=42, shuffle=True)
# Initialize classifiers
classifiers = {
    'Logistic Regression': LogisticRegression(max_iter=200),
    'Support Vector Machines': SVC(),
    'k-Nearest Neighbors': KNeighborsClassifier(),
    'Neural Network': MLPClassifier(max_iter=2000)
}

# Store results
results = []

# Perform 5-fold cross-validation for each model
for name, model in classifiers.items():
    scores = cross_val_score(model, X, y, cv=kfold, scoring='accuracy')
    results.append({
        'Model': name,
        'Mean Accuracy': scores.mean(),
        'Standard Deviation': scores.std()
    })

# Convert results to DataFrame for easy visualizations
results_df = pd.DataFrame(results)

# Display the results
print(results_df)


                     Model  Mean Accuracy  Standard Deviation
0      Logistic Regression       0.856559            0.066269
1  Support Vector Machines       0.928172            0.047601
2      k-Nearest Neighbors       0.927957            0.052440
3           Neural Network       0.935054            0.040466


## Results and Conclusion for Classification Experiments

From our experiments conducted on classification models, what we can conclude is that the best performing model is the Neural Network which has a mean accuracy of 0.94 or 94% as well as the lowest standard deviation of 0.02. This indicated that for our specific dataset in predicting CKD, Neural Networks are the best model which captures the underlying features in the dataset. The next most accurate models would be support vector machines as well as k-Nearest Neighbors. This may be a sign that more complex models are needed to provide accurate predictions for our datasets and using a simple logistic regression may not be sufficient in this case.

In [104]:
# Experiments with Neural Network.


# Define different configurations for the neural network
# Here we just focus on the number of hidden neurons
nn_configurations = [
    {"hidden_layer_sizes": (5,)},   # Single hidden layer with 5 neurons
    {"hidden_layer_sizes": (10,)},  # Single hidden layer with 10 neurons
    {"hidden_layer_sizes": (15,)},  # Single hidden layer with 15 neurons
]

# Store results for each configuration
nn_results = []

# Train and evaluate each configuration using 5-fold cross-validation
for config in nn_configurations:
    nn_model = MLPClassifier(**config, max_iter=2000)  # Use the current config
    nn_scores = cross_val_score(nn_model, X, y, cv=kfold, scoring='accuracy')
    nn_results.append({
        'Configuration': f"Hidden Neurons: {config['hidden_layer_sizes'][0]}",
        'Mean Accuracy': nn_scores.mean(),
        'Standard Deviation': nn_scores.std()
    })

# Convert results to DataFrame for easy visualization
nn_results_df = pd.DataFrame(nn_results)

# Display results of neural network configurations
print(nn_results_df)

        Configuration  Mean Accuracy  Standard Deviation
0   Hidden Neurons: 5       0.934839            0.035151
1  Hidden Neurons: 10       0.921935            0.047970
2  Hidden Neurons: 15       0.928387            0.047183


## Results and Conclusion for Neural Network Experiments

After running experiments with neural networks with varying numbers of hidden neurons, what we got for outputs is that the configuration with 5 neurons is the most effective model for mean score with a mean accuracy of 0.93 or 93%. However, the differences between mean accuracies between the model with 5 hidden neurons versus 15 or 10 weren't drastic, all providing good accuracy score. As for the standard deviations, a different pattern follows, with the model with 5 hidden neurons giving the lowest value of 0.05. Similar to the mean, the values for the standard deviation didn't vary too much between the models with 10 or 15 neurons either. What we can conclude from these results overall is that the best model from these experiments is the one with 5 hidden neurons which provided the best mean accuracy. With the limited tests, it is difficult to narrow down why as to the reason the model with lesser hidden neurons is the most effective. In order to effectively make predictions on which configurations would be optimal, more tests would need to be conducted. 